File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/validation/RegexValidatorTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

import org.junit.jupiter.api.Test;

class RegexValidatorTest {

    @Test
    void eval_ShouldReturnTrueForMatchingPattern() {
        RegexValidator validator = new RegexValidator("^[A-Z]+$");
        assertTrue(validator.eval("ABC"), "All uppercase letters should match");
    }

    @Test
    void eval_ShouldReturnFalseForNonMatchingPattern() {
        RegexValidator validator = new RegexValidator("^[A-Z]+$");
        assertFalse(validator.eval("abc"), "Lowercase letters should not match");
        assertFalse(validator.eval("123"), "Numbers should not match");
    }

    @Test
    void eval_ShouldReturnFalseForNullValue() {
        RegexValidator validator = new RegexValidator(".*");
        assertFalse(validator.eval(null), "Null value should not match");
    }

    @Test
    void eval_ShouldReturnTrueForEmptyStringWithMatchingPattern() {
        RegexValidator validator = new RegexValidator("^$");
        assertTrue(validator.eval(""), "Empty string should match");
    }

    @Test
    void eval_ShouldReturnTrueForComplexPattern() {
        RegexValidator validator = new RegexValidator("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}$");
        assertTrue(validator.eval("test@email.com"), "Valid email should match");
        assertFalse(validator.eval("invalid-email"), "Invalid email should not match");
    }

    @Test
    void constructor_ShouldCompilePattern() {
        RegexValidator validator = new RegexValidator("test");
        assertNotNull(validator.getCompiledPattern(), "Pattern should be compiled");
        assertEquals("test", validator.getCompiledPattern().pattern(), "Pattern should be the same as input");
    }

    @Test
    void serialization_ShouldPreserveState() throws Exception {
        RegexValidator originalValidator = new RegexValidator("^[A-Z]+$");
        ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
        ObjectOutputStream out = new ObjectOutputStream(byteOut);
        out.writeObject(originalValidator);

        ByteArrayInputStream byteIn = new ByteArrayInputStream(byteOut.toByteArray());
        ObjectInputStream in = new ObjectInputStream(byteIn);
        RegexValidator deserializedValidator = (RegexValidator) in.readObject();

        assertEquals(originalValidator.getCompiledPattern().pattern(),
                deserializedValidator.getCompiledPattern().pattern(), "Compiled patterns should match");
    }
}

File: lib/python/opentoken/src/test/opentoken/attributes/validation/regex_validator_test.py
"""
Tests for RegexValidator.
"""

import pickle
import pytest

from opentoken.attributes.validation.regex_validator import RegexValidator


class TestRegexValidator:
    """Test cases for RegexValidator."""

    def test_eval_should_return_true_for_matching_pattern(self):
        """Test that eval returns True for matching patterns."""
        validator = RegexValidator(r"^[A-Z]+$")
        assert validator.eval("ABC") is True, "All uppercase letters should match"

    def test_eval_should_return_false_for_non_matching_pattern(self):
        """Test that eval returns False for non-matching patterns."""
        validator = RegexValidator(r"^[A-Z]+$")
        assert validator.eval("abc") is False, "Lowercase letters should not match"
        assert validator.eval("123") is False, "Numbers should not match"

    def test_eval_should_return_false_for_null_value(self):
        """Test that eval returns False for null (None) values."""
        validator = RegexValidator(r".*")
        assert validator.eval(None) is False, "Null value should not match"

    def test_eval_should_return_true_for_empty_string_with_matching_pattern(self):
        """Test that eval returns True for empty strings with matching patterns."""
        validator = RegexValidator(r"^$")
        assert validator.eval("") is True, "Empty string should match"

    def test_eval_should_return_true_for_complex_pattern(self):
        """Test that eval works for complex patterns."""
        validator = RegexValidator(r"^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,6}$")
        assert validator.eval("test@email.com") is True, "Valid email should match"
        assert validator.eval("invalid-email") is False, "Invalid email should not match"

    def test_constructor_should_compile_pattern(self):
        """Test that the constructor compiles the pattern."""
        validator = RegexValidator(r"test")
        assert validator.compiled_pattern is not None, "Pattern should be compiled"
        assert validator.compiled_pattern.pattern == "test", "Pattern should be the same as input"

    def test_serialization_should_preserve_state(self):
        """Test serialization and deserialization of the validator."""
        original_validator = RegexValidator(r"^[A-Z]+$")

        # Serialize the validator using pickle
        serialized_data = pickle.dumps(original_validator)

        # Deserialize the validator
        deserialized_validator = pickle.loads(serialized_data)

        # Test that both validators behave identically
        assert original_validator.compiled_pattern.pattern == deserialized_validator.compiled_pattern.pattern, (
            "Compiled patterns should match after deserialization"
        )